sstable/src/sstable_index.rs

```diff
@@ -139,51 +278,456 @@ impl SSTableIndexBuilder {
+        bit_packer.write(
+            (last_block_addr.byte_range.end as i64 - range_pred + range_shift) as u64,
+            range_start_nbits,
+            &mut self.buffer_addrs,
+        )?;
+        bit_packer.flush(&mut self.buffer_addrs)?;
+
+        self.block_addrs.clear();
+        Ok(())
+    }
+
+    fn write_block_meta(&mut self, block_addr: BlockAddr) -> io::Result<()> {
+        self.block_addrs.push(block_addr);
+        if self.block_addrs.len() >= STORE_BLOCK_LEN {
+            self.flush_block()?;
+        }
+        Ok(())
+    }
+
+    fn serialize<W: std::io::Write>(&mut self, wrt: &mut W) -> io::Result<()> {
+        if !self.block_addrs.is_empty() {
+            self.flush_block()?;
+        }
+        let len = self.buffer_block_metas.len() as u64;
+        len.serialize(wrt)?;
+        wrt.write_all(&self.buffer_block_metas)?;
+        wrt.write_all(&self.buffer_addrs)?;
+        Ok(())
+    }
+}
+
+fn find_best_slop(elements: impl Iterator<Item = (usize, u64)> + Clone) -> (u32, u8) {
+    let slop_iterator = elements.clone();
+    let derivation_iterator = elements;
+
+    let mut min_slop_idx = 1;
+    let mut min_slop_val = 0;
+    let mut min_slop = u32::MAX;
+    let mut max_slop_idx = 1;
+    let mut max_slop_val = 0;
+    let mut max_slop = 0;
+    for (index, value) in slop_iterator {
+        let slop = (value / index as u64) as u32;
+        if slop <= min_slop {
+            min_slop = slop;
+            min_slop_idx = index;
+            min_slop_val = value;
+        }
+        if slop >= max_slop {
+            max_slop = slop;
+            max_slop_idx = index;
+            max_slop_val = value;
+        }
+    }
 
-    type ValueWriter = crate::value::index::IndexValueWriter;
+    // above is an heristic giving the "highest" and "lowest" point. It's imperfect in that in that
+    // a point that appear earlier might have a high slop derivation, but a smaller absolute
+    // derivation than a latter point.
+    // The actual best values can be obtained by using the symplex method, but the improvement is
+    // likely minimal, and computation is way more complexe.
+    //
+    // Assuming these point are the furthest up and down, we find the slop that would cause the same
+    // positive derivation for the highest as negative derivation for the lowest.
+    // A is the optimal slop. B is the derivation to the guess
+    //
+    // 0 = min_slop_val - min_slop_idx * A - B
+    // 0 = max_slop_val - max_slop_idx * A + B
+    //
+    // 0 = min_slop_val + max_slop_val - (min_slop_idx + max_slop_idx) * A
+    // (min_slop_val + max_slop_val) / (min_slop_idx + max_slop_idx) = A
+    //
+    // we actually add some correcting factor to have proper rounding, not truncation.
+
+    let denominator = (min_slop_idx + max_slop_idx) as u64;
+    let final_slop = ((min_slop_val + max_slop_val + denominator / 2) / denominator) as u32;
```

Comment:
```suggestion
    let final_slop = ((min_slop_val + max_slop_val + denominator / 2) / denominator) as u32;
```

would using a decimal slope (not a float but just using a 4 bits as decimal for instance) help here?

`let final_slop = ((min_slop_val + max_slop_val + denominator / 2) * 32 / denominator) as u32;`

Reply:
that code is enough to make sure we round in the right direction. I *think* using fixed point integers wouldn't really help as we would still need to use a similar trick when collapsing the fixed point integers to whole numbers

Reply:
I meant: wouldn't it make the decimal slope make the prediction for high indexes and eventually help with the compression.

> as we would still need to use a similar trick when collapsing the fixed point integers to whole numbers

I suspect you forgot there is a multiplication in between.

`truncateroundorwhatever(slope*idx)`
vs
`idx * truncateroundorwhatever(slope)`
will actually give you very different precision.

I would not be surprised if you could shave off one or two bits in your encoding.

Reply:
ftr: I tried and on my sample dataset, there was a negligible gain of about 0.1%. Either I did something wrong and I can't figure out what, or the impact is negligible